### ISSUE TITLE
kube-scheduler/controller-manager use non-unique identity in leader lock

### ISSUE THREAD

**Issue by anguslees** [kubernetes/kubernetes#61151]
/kind bug

Reopening https://github.com/kubernetes/kubernetes/issues/22775

**What happened**:

My HA controller-managers and schedulers are fighting over the same resources, even though I have `--leader-elect=true` set for both.

**What you expected to happen**:

Calmness and serenity.  In particular, these jobs should each elect a single leader and only that leader should operate on the cluster.

**How to reproduce it (as minimally and precisely as possible)**:

Run kube-scheduler/controller-manager in a (multi-host) `hostNetwork=true` daemonset, with `--leader-elect=true` and *the same system hostname on each node*.

The system hostname thing is because the nodes are booting from read-only system images, and don't have entries in DNS (or other external orchestration).  They're running a custom system image, but note (eg) CoreOS images will do similar out of the box.   The nodes *do* have unique kubernetes *node names* (of course).

**Anything else we need to know?**:

Note the `control-plane.alpha.kubernetes.io/leader` annotation of kube-controller-manager endpoint (eg) looks like:

```json
{
 "holderIdentity":"bananapi",
 "leaseDurationSeconds":15,
 "acquireTime":"2017-10-10T00:08:42Z",
 "renewTime":"2018-03-14T03:01:42Z",
 "leaderTransitions":0
}
```

The "holderIdentity" (`bananapi`) here is the system hostname, which is *not* unique, and I strongly suspect this is the cause of my issues.  

I strongly suggest we should use a holderIdentity value that is actually unique and tied to the identity of the pod.  Some suggestions: pod UID, a listening IP+port, a randomly generated in-memory token.

Fwiw, the fix attempted/abandoned in #22776 (change from hostNetwork=true to use hostPort instead) is not an option since the self-hosted controller-manager/scheduler (potentially) need to be scheduled before regular cluster pod networking is functional.  These _need_ to use hostNetwork=true.

**Environment**:
- Kubernetes version (use `kubectl version`):

Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.3", GitCommit:"f0efb3cb883751c5ffdbe6d515f3cb4fbe7b7acd", GitTreeState:"clean", BuildDate:"2017-11-08T18:27:48Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/arm"}

These are the `gcr.io/google_containers/kube-scheduler-arm:v1.8.3`, et al images.  I'd be very happy to hear this has already been fixed in more recent k8s versions.

- Cloud provider or hardware configuration: baremetal
- OS (e.g. from /etc/os-release): containos (a custom OS/image)
- Install tools: kubeadm self-hosted, plus manual HA post-install changes
- Others:

**Comment by liggitt** [comment]
fixed in 1.10 in 58302

**Comment by anguslees** [comment]
(woot)
